### PR TITLE
CMP: add support for genm/genp messages with `id-it-caCerts`

### PR DESCRIPTION
**Pull request by DDvO**
This pertains to the CMP API, the CMP app, the doc, and the tests.
Doing so, found and fixed a bug in `gen_new()` in `cmp_msg.c` checking wrong ITAVs.

The CMP client may send a `genm` message with infoType `id-it-caCerts` (using option `-infotype caCerts`)
requesting CA certs, which the server provides in a `genp` message also with infoType `id-it-caCerts`.
The client performs sanity checks on the obtained certs and saves them in file given with `-cacertsout`.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/cmp/cmp_genm.c
Comment (t8m) [inline crypto/cmp/cmp_genm.c:70]: Nit: IMO && should be aligned with ossl_...

File: crypto/cmp/cmp_genm.c
Comment (t8m) [inline crypto/cmp/cmp_genm.c:98]: Typo: TOOD

File: doc/man3/OSSL_CMP_exec_certreq.pod
Comment (t8m) [inline doc/man3/OSSL_CMP_exec_certreq.pod:172]: OpenSSL 3.2 here

File: include/openssl/cmp.h.in
Comment (t8m) [inline include/openssl/cmp.h.in:143]: Why this dropped line?

File: include/openssl/cmp.h.in
Comment (t8m) [inline include/openssl/cmp.h.in:206]: Why this?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/cmp/cmp_genm.c
Comment (DDvO) [reply crypto/cmp/cmp_genm.c:70]: Thanks for your comments!

`check-format.pl crypto/cmp/cmp_genm.c`, as well as the Emacs C mode, is happy with this indentation.
The idea is that on a multi-line assignment (and in similar other cases),
indentation is increased by the usual delta (+4), which is independent of the length of the variable name on the LHS.

File: crypto/cmp/cmp_genm.c
Comment (DDvO) [reply crypto/cmp/cmp_genm.c]: Oops, fixed.

File: doc/man3/OSSL_CMP_exec_certreq.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_exec_certreq.pod]: Updated.

File: include/openssl/cmp.h.in
Comment (DDvO) [reply include/openssl/cmp.h.in:143]: I deleted the empty line to make clear that the following decl
```
typedef ASN1_BIT_STRING OSSL_CMP_PKIFAILUREINFO;
```
is directly related.

File: include/openssl/cmp.h.in
Comment (DDvO) [reply include/openssl/cmp.h.in:206]: Same reason as above.

**Comment by DDvO** [comment]
@t8m ok now?

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Asking @openssl/committers for 2nd review.

**Comment by DDvO** [comment]
Rebased to fix (trivial) merge conflict in `libcrypto.num`.

Ping @openssl/committers for 2nd review.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Rebased again to fix trivial merge conflicts.

Asking again @openssl/committers for 2nd review.

**Comment by DDvO** [comment]
Can we please get this done now?
2nd approval is missing.

**Comment by DDvO** [comment]
Fixed clashes of two `CMP_R` codes that came in when rebasing.

**Review by t8m — APPROVED** [review]
Still OK

**Comment by DDvO** [comment]
@paulidale can you do the 2nd review/approval?

File: apps/cmp.c
Comment (paulidale) [inline apps/cmp.c:2830]: #19205 has been merged.

File: apps/cmp.c
Comment (paulidale) [inline apps/cmp.c:2800]: We try to avoid littering the code with TODOs...

File: crypto/cmp/cmp_asn.c
Comment (paulidale) [inline crypto/cmp/cmp_asn.c:214]: indentation?

File: crypto/cmp/cmp_genm.c
Comment (paulidale) [inline crypto/cmp/cmp_genm.c:97]: This has been merged.

File: crypto/cmp/cmp_genm.c
Comment (paulidale) [reply crypto/cmp/cmp_genm.c:70]: I'd have indented this as @t8m suggests

**Review by paulidale — COMMENTED** [review]
A fairly superficial review, I know nothing about what this is doing...

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Good that you spotted this - resolved this TODO, as well as the two further ones referring to #19205.

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Re-phrased.

File: crypto/cmp/cmp_asn.c
Comment (DDvO) [reply crypto/cmp/cmp_asn.c]: Well, indentation is actually correct here: +4 in multi-line assignments (like also in other similar cases).
I've just placed the `?` on the line before now, which should improve the appearance.

File: crypto/cmp/cmp_genm.c
Comment (DDvO) [reply crypto/cmp/cmp_genm.c]: I've reworked the assignment to generally make this code more readable.
As a by-product, the indentation issue disappears.

**Comment by DDvO** [comment]
> A fairly superficial review, I know nothing about what this is doing...

Thank you @paulidale for anyway doing it!
I've just addressed all comments.

**Comment by DDvO** [comment]
@paulidale can you approve now, given the above changes?

**Comment by DDvO** [comment]
The two CI failures are unrelated.

**Comment by DDvO** [comment]
Rebased to fix a merge conflict.

**Comment by DDvO** [comment]
Pinging again @paulidale - ok now?

File: crypto/cmp/cmp_genm.c
Comment (paulidale) [reply crypto/cmp/cmp_genm.c:97]: This is still here, either the code or comment is wrong.

**Comment by paulidale** [comment]
@openssl/committers, looking for a second review some someone with half a clue here.

File: apps/cmp.c
Comment (vdukhovni) [inline apps/cmp.c:2810]: Is there really no suitable `opt_infotype` to check for in this branch?  An explicit choice feels safer than a default action.

File: crypto/cmp/cmp_genm.c
Comment (vdukhovni) [inline crypto/cmp/cmp_genm.c:50]: What is the expected result if the `type_CA > 1`? Would `!= 0` be better here?

File: crypto/cmp/cmp_genm.c
Comment (vdukhovni) [inline crypto/cmp/cmp_genm.c:88]: A less cryptic error message here?

File: crypto/cmp/cmp_genm.c
Comment (vdukhovni) [inline crypto/cmp/cmp_genm.c:109]: It looks like `ITAV` is not a standard term in the context of CMP, but rather is OpenSSL-specific jargon.  While it may be reasonable to use in API types, ... where developers will have read the documentation, it should probably not leak into error messages seen by users.

If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Changed to "CA certificates".

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Changed to `CMP_err1("Failed to create genm for -infotype %s", opt_infotype_s);`

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Changed to "Did not receive response on genm or genp is not valid".

File: crypto/cmp/cmp_genm.c
Comment (DDvO) [reply crypto/cmp/cmp_genm.c]: Re-phrased to "response on genm requesting infoType %s does not include suitable value".

> If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

Most OpenSSL lib error messages are rather cryptic and give pretty little detail such that they are typically hard to make use of, partly even for experts.
When I put error messages like in this PR, I try to be to-the-point and tend to give more detail, 
which hopefully will be help experts, while regular users anyway will not get much out of OpenSSL error queue dumps.

OTOH, the diagnostics of the CMP app should be digestible for regular users,
and with the improvement you suggested above I believe this has been achieved for this PR.

**Comment by DDvO** [comment]
Thank you @vdukhovni for your review comments.
I've meanwhile addressed all of them.

**Review by vdukhovni — COMMENTED** [review]
I'll leave it to others to approve the PR, and gauge whether the updated error messages are now sufficiently user-friendly. I have no further issues to raise.

**Comment by DDvO** [comment]
> @openssl/committers, looking for a second review from someone with half a clue here.

> I'll leave it to others to approve the PR, 
and gauge whether the updated error messages are now sufficiently user-friendly. 
I have no further issues to raise.

@paulidale can you approve now, since @vdukhovni is fine with the technical contents?

**Comment by paulidale** [comment]
@vdukhovni has approval permission...

**Comment by DDvO** [comment]
> @vdukhovni has approval permission...

@vdukhovni, would you mind approving yourself?
@paulidale appears fine with your remaining concerns, namely the user-friendliness of the (updated) error messages.

**Comment by DDvO** [comment]
Since two weeks we are in a reviewer approval deadlock on this PR.
Can any of the @openssl/committers please help out.

File: apps/cmp.c
Comment (t8m) [inline apps/cmp.c:2837]: Typo: provided

**Review by t8m — CHANGES_REQUESTED** [review]

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: oops, fixed

**Comment by DDvO** [comment]
Rebased to fix merge conflicts

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
@openssl/committers, can we please get the 2nd approval done now?

**Review by paulidale — APPROVED** [review]

**Comment by DDvO** [comment]
Thanks a lot @paulidale for jumping in :tada:

**Comment by DDvO** [comment]
Merged - thanks @t8m and @paulidale for your comments and approvals and @vdukhovni for your comments.

**Pull request closed without merging**
